tests/test_autofocus.py

```diff
@@ -0,0 +1,113 @@
+    (NO_CH_ZSTACK, ("c",), ()),
+    (TWO_CH_ZSTACK, ("c",), (0, 4)),
+    (TWO_CH_ZSTACK, ("z",), range(0, 11, 2)),
+    (TWO_CH_ZSTACK.replace(axis_order="tpgzc"), ("c",), range(0, 11, 2)),
+    (TWO_CH_ZSTACK.replace(axis_order="tpgzc"), ("z",), (0, 3, 6)),
+    (TWO_CH_ZSTACK, ("z",), range(0, 11, 2)),
+    (TWO_CH.replace(grid_plan=GRID_PLAN), ("g",), (0, 3)),
+    (TWO_CH_TWO_P, ("g",), ()),
+    (TWO_CH_TWO_P_GRID, ("g",), range(0, 10, 3)),
+    (TWO_CH_TWO_P_GRID, ("p",), (0, 5)),
+    (TWO_CH_TWO_P_T, ("t",), (0, 5)),
+    (TWO_CH_TWO_P_T, ("t", "p"), range(0, 10, 3)),
+    (TWO_CH_SUBPAF_C, (), (2, 4)),
+    (TWO_CH.replace(stage_positions=[ZPOS_30, SUB_P_AF_G]), (), ()),
+    (TWO_CH.replace(stage_positions=[ZPOS_30, SUB_P_AF_G], grid_plan=GRID_PLAN), (), (4, 7)),
+    (TWO_CH.replace(stage_positions=[ZPOS_30, SUB_P_AF_P], grid_plan=GRID_PLAN), (), (4,)),
+    (TWO_CH_SUBPAF_C.replace(grid_plan=GRID_PLAN), (), range(4, 11, 2)),
+    (TWO_CH_SUBPAF_C.replace(z_plan=ZRANGE2), (), (6, 10)),
+    (TWO_CH_SUBPAF_Z.replace(z_plan=ZRANGE2), (), range(6, 17, 2)),
+    (TWO_CH_SUBPAF_Z.replace(z_plan=ZRANGE2), ("p",), (0, *tuple(range(7, 18, 2)))),
+    (TWO_CH_SUBPAF_C.replace(z_plan=ZRANGE2, grid_plan=GRID_PLAN), ("c",), range(0, 29, 4)),
+    (TWO_CH.replace(stage_positions=[ZPOS_30, useq.Position(z=10, sequence=MDASequence(autofocus_plan=AF_C, grid_plan=GRID_PLAN))]), (), (2, 5)),
+    (TWO_CH.replace(stage_positions=[SUB_P_AF_C, useq.Position(z=10, sequence=MDASequence(autofocus_plan=AF_G, grid_plan=GRID_PLAN))]), (), range(0, 11, 2)),
+]
+# fmt: on
+
+
+@pytest.mark.parametrize("mda, af_axes, expected_af_indices", AF_TESTS)
+def test_autofocus(
+    mda: MDASequence,
+    af_axes: tuple[str, ...],
+    expected_af_indices: Iterable[int],
+) -> None:
+    if af_axes:
+        mda = mda.replace(autofocus_plan=AF.copy(update={"axes": af_axes}))
+        assert isinstance(mda.autofocus_plan, AxesBasedAF)
+        assert mda.autofocus_plan.axes == af_axes
+
+    actual = [i for i, e in enumerate(mda) if isinstance(e.action, HardwareAutofocus)]
+    expected = list(expected_af_indices)
+    if expected != actual:
+        raise AssertionError(f"Expected AF indices at {expected} but got {actual}")
+
+
+def test_autofocus_z_pos() -> None:
+    mda = TWO_CH.replace(
+        stage_positions=[ZPOS_200], z_plan=ZRANGE2, autofocus_plan=AF_C
+    )
+    assert all(e.z_pos == 200 for e in mda if isinstance(e.action, HardwareAutofocus))
+
+
+def test_autofocus_z_pos_abovebelow() -> None:
+    mda = TWO_CH.replace(
+        stage_positions=[ZPOS_200],
+        z_plan=useq.ZAboveBelow(above=2, below=2, step=1),
+        autofocus_plan=AF_C,
+    )
+    assert all(e.z_pos == 200 for e in mda if isinstance(e.action, HardwareAutofocus))
+
+
+def test_autofocus_z_pos_af_sub_sequence() -> None:
+    mda = TWO_CH.replace(stage_positions=[SUB_P_AF_C], z_plan=ZRANGE2)
+    assert all(e.z_pos == 10 for e in mda if isinstance(e.action, HardwareAutofocus))
```

Comment:
quick note @fdrgsp  ... note how this is much more interpretable to a reader:

```python
assert all(e.z_pos == 10 for e in mda if isinstance(e.action, HardwareAutofocus))
```
than this:

```python
z = [e.z_pos for e in mda]
assert z[2] == z[4] == 10
```